erica/infrastructure/sqlalchemy/database.py

```diff
@@ -20,14 +20,14 @@ def orjson_deserializer(json):
 
 def get_engine():
     return create_engine(
-        get_settings().database_url, json_serializer=orjson_serializer, json_deserializer=orjson_deserializer)
+        get_settings().database_url, **engine_args)
 
 
 class DatabaseSessionProvider(Provider[Session]):
 
     def get(self) -> Session:
-        engine = get_engine()
-        if not database_exists(engine.url):
-            create_database(engine.url)
-        SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)
-        return SessionLocal()
+        return db.session
```

Comment:
It gets confusing quickly, so let's differentiate and define some terms:
- a `sqlalchemy.orm.Session` instance (the "**session**") - used to interact with the database
- a `sqlachemy.orm.sessionmaker` instance (the "**sessionmaker**" or "session class") - can be called to create a session

Before my changes, when asked by opyoid, the DatabaseSessionProvider created a **sessionmaker** (`SessionLocal`) and called it to provide a **session** instance (`SessionLocal()`) to the JobService.

That was problematic because we somehow need to get a reference to _that exact session_ object when FastAPI finishes a request context and also when a worker job has finished or failed. If opyoid has a way to do that (in a way that's asyncio-context-safe), then I haven't found it.

After my changes, it still returns a **session** instance (`db.session`), however it no longer creates it nor does it create a **sessionmaker**. Both are created fastapi_sqlachemy.

For using something in a `with` block, you need a context manager (an object with `__enter__()` and `__exit__()` methods) - neither sessions nor sessionmakers are context managers. The `fastapi_sqlalchemy.db` object is [a custom context manager](https://github.com/mfreeborn/fastapi-sqlalchemy/blob/master/fastapi_sqlalchemy/middleware.py#L64) that handles session creation and teardown.

Does that help?

Reply:
Thank you Niko :raised_hands: I got it :smile: